src/main/java/ro/developmentfactory/thedrones/service/DronePositionServiceImpl.java

```diff
@@ -0,0 +1,146 @@
+package ro.developmentfactory.thedrones.service;
+
+import jakarta.persistence.EntityNotFoundException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import ro.developmentfactory.thedrones.controller.dto.DroneStatusResponse;
+import ro.developmentfactory.thedrones.repository.DroneRepository;
+import ro.developmentfactory.thedrones.repository.DroneStatusRepository;
+import ro.developmentfactory.thedrones.repository.entity.Direction;
+import ro.developmentfactory.thedrones.repository.entity.Drone;
+import ro.developmentfactory.thedrones.repository.entity.DroneStatus;
+
+import java.util.UUID;
+
+@Service
+public class DronePositionServiceImpl implements DronePositionService {
+
+    private final DroneStatusRepository droneStatusRepository;
+    private final DroneRepository droneRepository;
+    private static final Logger log = LoggerFactory.getLogger(DronePositionServiceImpl.class);
+
+    public DronePositionServiceImpl(DroneStatusRepository droneStatusRepository, DroneRepository droneRepository) {
+        this.droneStatusRepository = droneStatusRepository;
+        this.droneRepository = droneRepository;
+    }
+
+    @Override
+    @Transactional
+    public DroneStatusResponse turningDirection(DroneStatusResponse droneStatusResponse) {
+        UUID idDrone = droneStatusResponse.getIdDrone();
+
+        Drone drone = droneRepository.findById(idDrone)
+                .orElseThrow(() -> new EntityNotFoundException("Drone with id " + idDrone + " not found"));
+
+        if (droneStatusResponse.getTurnDirection() == null) {
+            throw new IllegalArgumentException("Turn direction cannot be null");
+        }
+
+        DroneStatus currentStatus = drone.getDroneStatus();
+
+        Direction newDirection;
+
+        if (droneStatusResponse.getTurnDirection() == DroneStatusResponse.TurnDirection.RIGHT) {
+            newDirection = currentStatus.getFacingDirection().turnRight();
+        } else {
+            newDirection = currentStatus.getFacingDirection().turnLeft();
+        }
+
+        currentStatus.setFacingDirection(newDirection);
+
+        droneStatusRepository.save(currentStatus);
+        return convertToResponse(currentStatus);
+    }
+
+    @Override
+    public DroneStatusResponse moveForward(DroneStatusResponse droneStatusResponse) {
+        UUID idDrone = droneStatusResponse.getIdDrone();
+        log.debug("Moving forward for drone: {}", idDrone);
+
+        Drone drone = droneRepository.findById(idDrone)
+                .orElseThrow(() -> new EntityNotFoundException("Drone with id " + idDrone + " not found"));
+
+        DroneStatus droneStatus = drone.getDroneStatus();
+        if (droneStatus == null) {
+            throw new IllegalArgumentException("DroneStatus must not be null");
+        }
```

Comment:
for which reason it can be null? only if there was an error during saving?
I'd say that you can assume it's not null, worst case you will get a null pointer exception that will be fought by an exception handler and will be translated to an internal server error, which is right. but if you want to have a validation like this, better use [Objects.requireNonNull(droneStatus)](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Objects.html#requireNonNull(T)), see more [here](https://stackoverflow.com/a/60712789/832095)